### DISCUSSION TITLE
Build Pinocchio with HPP-FCL support option

### DISCUSSION THREAD

**Discussion by sholybest** [stack-of-tasks/pinocchio#2223] [Q&A]
First, thank you to everyone reading this message.
When I ran a meshcat display code that shows a model from example-robot-data, I got some errors same with those in this link https://github.com/stack-of-tasks/pinocchio/issues/1371, "No to_python (by-value) converter found for C++ type ...".

As the answer suggested at the link, I compiled pinocciho adding hpp-fcl support option as following.
`cmake .. -DBUILD_WITH_COLLISION_SUPPORT=ON -DBUILD_WITH_OPENMP_SUPPORT=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-march=native" 
`

However, when `BUILD_WITH_COLLISION_SUPPORT` option was added, I got an error message as in the attached image below.
![image](https://github.com/stack-of-tasks/pinocchio/assets/68987433/4f5f4544-9a55-4f4a-85ac-683155baa533)

The versions of the dependencies are as follows. When I installed them with pip install or robotpkg, these same versions didn't make problems and displayed the robot well. How can I solve this error??
Eigen: 3.3.7
eigenpy: 3.4.0
Pinocchio: 2.7.0
hpp-fcl: 2.4.4
Boost: 1.83.0

Thank you.

**Comment by jorisv** [comment]
Hello @sholybest,

Has I remember, this is an issue between boost serialization and Eigen.
It has been fixed in this [hpp-fcl PR](https://github.com/humanoid-path-planner/hpp-fcl/pull/530/files).

Unfortunately, this patch is not released. We will release the new hpp-fcl version soon, but in the mean time, you can build hpp-fcl from source on the following commit : 40a636b40e6ca3caf27b3d580de0697044c406af

**Reply by sholybest** [reply]
Thank you very much for your answer.
I uninstalled and reinstalled hpp-fcl and pinocchio. 
hpp-fcl was built from the suggested commit, but when building pinocchio, the same error still remains..
Could you give me some additional advices?

This is the error message from pinocchio this time.
![image](https://github.com/stack-of-tasks/pinocchio/assets/68987433/a716fad9-4dc6-4b90-ac06-0cce9f7c04ba)


p.s.
I got an access to the given commit as in the below image. Is this a right way to build hpp-fcl?
![image](https://github.com/stack-of-tasks/pinocchio/assets/68987433/56ee6ea6-955a-43b4-bb1e-12885cb19ced)

Thank you.

**Reply by jorisv** [reply]
I can't reproduce your issue on my computer.

When building pinocchio 2.7.0 with:

- Eigen: 3.3.7
- eigenpy: 3.4.0
- hpp-fcl: 2.4.4

Every thing is working fine.
What compiler are you using ? On what OS are you ?
Can you run `cmake . -DCMAKE_EXPORT_COMPILE_COMMANDS=ON` in your build directory and send me the compile_commands.json file generated ?

**Reply by sholybest** [reply]
I'm working on Ubuntu20.04 and gcc 11.4.0 compiler.
Here is the compile_commands.json file.
Thank you for your consideration..!

[compile_commands.json](https://github.com/stack-of-tasks/pinocchio/files/15074983/compile_commands.json)

**Reply by jorisv** [reply]
@sholybest It's the right way to build hpp-fcl, but did you install it after `make install` ? And did you install it in the right directory ?

I'm not able to look at your compilation error picture anymore. Can you publish them again ?

What branch/tag do you use to build pinocchio ?

**Reply by sholybest** [reply]
Yes, I've run `make install` for hpp-fcl and found "libhpp-fcl.so" at `/usr/local/lib.`
The previous image was not available for me, but I could see that when I refresh the page. Here is the image.
I'm working with the default master branch of pinocchio. Wasn't it a good choice?

![image](https://github.com/stack-of-tasks/pinocchio/assets/68987433/dc8ece57-5313-4af9-b5cb-c6855b7c73d2)


p.s.
I tried with devel branch, but got the same error. This is the message with devel branch.

![image](https://github.com/stack-of-tasks/pinocchio/assets/68987433/2e72a9a4-1ecc-476e-ad11-382cc4aea792)

**Reply by jorisv** [reply]
@sholybest master branch is the good choice. It's just weird I can't reproduce on my computer.

In your compile_commands.json, I don't see `/usr/local/include` as include directory. So, if you have installed `hpp-fcl` in `/usr/local/` the compiler don't see it.
You need to install `hpp-fcl` in the directory you have installed the old one. Maybe `/home/nam/.local/lib/python3.8/site-packages/cmeel.prefix/`.

**Reply by jorisv** [reply]
Another option is to install pinocchio with conda. It has hpp-fcl support.

- [install conda](https://docs.anaconda.com/free/miniconda/)
- [create and activate a conda virtualenv](https://docs.anaconda.com/free/working-with-conda/configurations/switch-environment/)
- install pinocchio in it: `conda install pinocchio -c conda-forge`
- install meshcat-python: `conda install meshcat-python -c conda-forge`

Then, you can use pinocchio in the conda virtualenv you created.

**Reply by sholybest** [reply]
I wanted to enhance the performance of the whole project, by compiling all the dependencies with appropriate options for my native architecture, rather than getting pre-compiled packages. 

I expected cmake to automatically find packages in `/usr/local` directory.. 
After getting your advice, I set CMAKE_PREFIX_PATH environment variable, and added redundantly in CmakeList.txt the followings:
`include_directories(/usr/local/include)`
`link_directories(/usr/local/lib)`
`link_libraries(hpp-fcl)`

Then cmake found `/usr/local/libhpp-fcl.so` as in the image below. 
![image](https://github.com/stack-of-tasks/pinocchio/assets/68987433/7fe04d2f-34b0-4cda-b08d-5b2bd8795c42)

But still same error remains..
When I pip installed hpp-fcl (at `/home/nam/.local/lib/python3.8/site-packages/cmeel.prefix/`), the error was removed and pinocchio could be built with hpp-fcl support option. The robot model could be successfully displayed in meshcat.
It did not make a big difference in computation time, so I would be satisfied with this option.

Maybe my system is messed up during many trials.. Sorry for bothering you with this unskilled question, and I appreciate your consideration and patience. I'll looking forward the new releases!

**Reply by nim65s** [reply]
You probably need to isolate your build process from anything else. If you have multiple setups configured at the same time, or some setup in a standard location (like /usr or /usr/local), those will prevent you from correctly configuring, compiling and running your projects.

Maybe you can try to do that in a Dockerfile, to ensure isolation and reproducibility ? Otherwise, can you share your full clone / configure / build process with a script, so we can review it ?

**Reply by sholybest** [reply]
Sorry for being late and that I don't have any experience with Dockerfiles.. but I can leave the procedure I did.

CPU: Intel i5-6600k
OS: Ubuntu20.04


1. Boost downloaded from https://www.boost.org/
```
tar xf boost_1_83_0.tar.bz2
cd boost_1_83_0

./bootstrap.sh --prefix=/usr/local --with-python=python3
sudo ./b2 -j2 toolset=gcc cxxflags="-O3 -march=native" threading=multi link=shared install 
```
2. urdfdom-headers
```
pip install cmeel-urdfdom-headers
```
3. eigenpy
Source code of 3.4.0 downloaded from https://github.com/stack-of-tasks/eigenpy/releases 
```
cd eigenpy 
mkdir build && cd build
cmake .. -DBUILD_WITH_ACCELERATE_SUPPORT=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-march=native 
-fopenmp" 
sudo make -j2 
sudo make install
```

4. hpp-fcl
```
git clone --recursive https://github.com/humanoid-path-planner/hpp-fcl
cd hpp-fcl
git reset --hard 40a636b40e6ca3caf27b3d580de0697044c406af
mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-march=native -fopenmp" 
sudo make -j2 
sudo make install
```

5. pinocchio
Added this in CmakeLists.txt
```
include_directories(/usr/local/include)
link_directories(/usr/local/lib)
link_libraries(hpp-fcl)
```
```
git clone --recursive https://github.com/stack-of-tasks/pinocchio.git
cd pinocchio
mkdir build && cd build
cmake .. -DBUILD_WITH_COLLISION_SUPPORT=ON -DBUILD_WITH_OPENMP_SUPPORT=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-march=native" 
sudo make -j2
sudo make install
```

**Reply by nim65s** [reply]
Thanks, this help.

First, I would like to ask what is the rationale for downloading boost instead of using eq. libboost-all-dev ubuntu package and not taking the latest release ? and cmeel-urdfdom-headers  instead of liburdfdom-headers-dev ubuntu package ? And not taking the latest eigenpy release ?

Also, I would say that if you are seeking better performances, I think upgrading your CPU is a way simpler option than trying to build everything with march=native by yourself. 6600K is nice, but it is almost 10 years old.

**Reply by sholybest** [reply]
Thank you very much for your concerns and supports. 
I wanted to do the real-time MPC simulation of quadruped with Box-FDDP solver of Crocoddyl library. When I pip installed Crocoddyl, the same versions mentioned above were installed automatically. I just followed them and there's no other reasons for choosing the versions. I didn't know about urdfdom-headers at all, but I could get it by pip install. That's all..

The SBC implemented in our quadruped robot platform is very old one, and its performance is similar to that of my desktop. This spec was sufficient for our previous MPC controller, with Single Rigid Body model. So I wanted to see even the most simple locomotion code of Crocoddyl would not be affordable. I would try my best and discuss with my team members and advisor, to upgrade the CPU. Do you think this spec is not enough for full-body DDP of quadruped?

**Comment by YuehChuan** [comment]
@sholybest  @nim65s thanks for the thread I wonder the issue is resolve or not. 
I get same issue using pink library, and I build pinnochio from source, here How I did
https://github.com/stephane-caron/pink/issues/179
hpp-fcl need to compile or using coal, thnaks.

ubuntu 24.04
Python: 3.12.3 [GCC 13.3.0]
  
casadi version    : 3.6.7
pinocchio version : 3.1.0
eigenpy version   : 3.8.0


environment.sh

```bash=
export /home/user/software/python3/dist-package/lib/python3.12/dist-packages:$PYTHONPATH
export VIRTUAL_ENV=/home/user/software/python3/dist-package
export CASADI_PREFIX=/home/user/software/python3/dist-package
```

**eigenpy-3.8.0.**

```bash=
tar -xzf eigenpy-3.8.0.tar.gz
cd eigenpy3.8.0
mkdir build && cd build
cmake .. \
  -DCMAKE_BUILD_TYPE=Release \
-DCMAKE_INSTALL_PREFIX=~/software/python3/dist-package \
  -DPYTHON_EXECUTABLE=$(which python3)  

make -j4
make install
```

unzip casadi-3.6.7.zip
mkdir build && cd build

sudo apt update
sudo apt install -y coinor-libipopt-dev
sudo apt install swig liblapack-dev libblas-dev -y

**casadi-3.6.7**
```bash=
 cmake .. -DCMAKE_BUILD_TYPE=Release \
      -DPYTHON_EXECUTABLE=$(which python3) \
      -DCMAKE_INSTALL_PREFIX=$VIRTUAL_ENV \
      -DWITH_IPOPT=ON \
      -DWITH_PYTHON=ON

  make -j4
  sudo make install
```
  **pinocchio-3.1.0**
```bash=
tar -xzf  pinocchio-3.1.0.tar.gz

mkdir build && cd build

    cmake .. \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_CXX_FLAGS="-I${VIRTUAL_ENV}/include -L${VIRTUAL_ENV}/lib" \
    -DCMAKE_INSTALL_PREFIX="$VIRTUAL_ENV" \
    -DPYTHON_EXECUTABLE=$(which python3) \
    -DBUILD_WITH_CASADI_SUPPORT=ON \
    -DCMAKE_PREFIX_PATH="$VIRTUAL_ENV" \
    -DCMAKE_INCLUDE_PATH="${VIRTUAL_ENV}/include" \
    -Dcasadi_DIR="${VIRTUAL_ENV}/lib/cmake/casadi" \
    -DCMAKE_LIBRARY_PATH="${VIRTUAL_ENV}/lib"

make -j4
make install
```

check_pinocchio_cpin.py install with python script
```python=
#!/usr/bin/env python3
import sys
print("Python:", sys.version)

print("\n[1] import modules")
import casadi as ca
import pinocchio as pin
from pinocchio import casadi as cpin
from pinocchio.visualize import MeshcatVisualizer
import eigenpy

print("  casadi version    :", ca.__version__)
print("  pinocchio version :", pin.__version__)
print("  eigenpy version   :", eigenpy.__version__)

print("\n🎉 ALL CHECKS PASSED — Pinocchio + cpin + CasADi WORKS")
```

https://github.com/stack-of-tasks/pinocchio/pull/2288

<img width="484" height="580" alt="Screenshot from 2026-01-23 13-46-46" src="https://github.com/user-attachments/assets/2fd9bc25-f72d-4013-986e-d2759bcf1f0e" />

**Reply by nim65s** [reply]
Can you explain first what is your goal, and second what exactly is your issue ?
The reproduction steps are welcome, but only if we can understand the first 2 parts.

**Reply by YuehChuan** [reply]
Sorry for late reply.
My goal is to use ROS noetic with pinocchio.
My issue is conda integration with ROS.
So I tried to compiled pinnochio,  casadi from source in ubuntu 24.04.
I can use pinnochio, but the visualization part with meshcat is broke.
THe error shows

" No to_python (by-value) converter found for C++ type ...."

Same as the issue above.
Thanks.

**Reply by nim65s** [reply]
Ok, to address this issue, can we ignore the ROS noetic, conda, and pip parts, and just ensure that pinocchio is working for you on ubuntu 24.04 ?
do you really need casadi ?

**Reply by YuehChuan** [reply]
Build from source, pinocchio,casadi works on ubuntu24.04. But the visualization parts using meshcat not work.
casadi used here  https://github.com/unitreerobotics/xr_teleoperate/blob/main/teleop/robot_control/robot_arm_ik.py#L6
but I can change use other ik.
Thanks.

**Reply by nim65s** [reply]
what do you mean by "meshcat not work" ? do you have an error message ?
